src/libPMacc/include/math/complex/Complex.tpp

```diff
@@ -0,0 +1,203 @@
+#include "algorithms/TypeCast.hpp"
+ 
+namespace PMacc
+{
+namespace algorithms
+{
+namespace math
+{
+
+/*  Set primary template and subsequent specialization for returning a complex number by using Euler's formula. */
+
+template<typename T_Type>
+struct Euler;
+
+template<typename T_Type>
+HDINLINE typename Euler< T_Type >::result euler(const T_Type& magnitude, const T_Type& phase)
+{
+    return Euler< T_Type > ()(magnitude, phase);
+}
+
+template<typename T_Type>
+HDINLINE typename Euler< T_Type >::result euler(const T_Type& magnitude, const T_Type& sinValue, const T_Type& cosValue)
+{
+    return Euler< T_Type > ()(magnitude, sinValue, cosValue);
+}
+
+template<typename T_Type>
+struct Euler
+{
+    typedef typename ::PMacc::math::Complex<T_Type> result;
+    
+    HDINLINE result operator( )(const T_Type &magnitude, const T_Type &phase)
+    {
+        return result(magnitude * PMacc::algorithms::math::cos(phase),magnitude * PMacc::algorithms::math::sin(phase));
+    }
+    
+    HDINLINE result operator( )(const T_Type &magnitude, const T_Type &sinValue, const T_Type &cosValue)
+    {
+        return result(magnitude * cosValue, magnitude * sinValue);
+    }
+};
+
+/* Specialize sqrt() for complex numbers. */
+
+template<typename T_Type>
+struct Sqrt< ::PMacc::math::Complex<T_Type> >
+{
+    typedef typename ::PMacc::math::Complex<T_Type> result;
+    
+    HDINLINE result operator( )(const ::PMacc::math::Complex<T_Type>& other)
+    {
+        if (other.get_real()<=0.0 && other.get_imag()==0.0) {
+            return ::PMacc::math::Complex<T_Type>(0.0, PMacc::algorithms::math::sqrt( -other.get_real() ) );
+        }
+        else {
+            return PMacc::algorithms::math::sqrt( PMacc::algorithms::math::abs(other) )*(other+PMacc::algorithms::math::abs(other))
+                /PMacc::algorithms::math::abs(other+PMacc::algorithms::math::abs(other));
+        }
+    }
+};
+
+/* Specialize exp() for complex numbers. */
+
+template<typename T_Type>
+struct Exp< ::PMacc::math::Complex<T_Type> >
+{
+    typedef typename ::PMacc::math::Complex<T_Type> result;
+    
+    HDINLINE result operator( )(const ::PMacc::math::Complex<T_Type>& other)
+    {
+        return PMacc::algorithms::math::euler(1.0,other.get_imag())*PMacc::algorithms::math::exp(other.get_real());
+    }
+};
+
+/*  Set primary template and subsequent specialization of arg() for retrieving
+ *  the phase of a complex number (Note: Branchcut running from -infinity to 0).
+ */
+template<typename T_Type>
+struct Arg;
+
+template<typename T_Type>
+HDINLINE typename Arg< T_Type >::result arg(const T_Type& val)
+{
+    return Arg< T_Type > ()(val);
+}
+
+template<typename T_Type>
+struct Arg< ::PMacc::math::Complex<T_Type> >
+{
+    typedef typename ::PMacc::math::Complex<T_Type>::type result;
+    
+    HDINLINE result operator( )(const ::PMacc::math::Complex<T_Type>& other)
+    {
+        if (other.get_real()==0.0 && other.get_imag()==0.0) return 0.0;
```

Comment:
please add a cast for static numbers like `0.0` -> should be type save `T_Type(0.0)`
re factoring suggestion:
```C++
typedef T_Type type;
if ( other.get_real() == type(0.0) )
{
  if( other.get_imag() == type(0.0)  ) return type(0.0);
   else if ( other.get_imag() > type(0.0) ) return type(M_PI)/type(2.0);
   else if ( other.get_imag() < type(0.0) ) return type(-M_PI)/type(2.0);
}
else 
if(other.get_real() < type(0.0) && other.get_imag() == type(0.0) ) return type(M_PI);

return PMacc::algorithms::math::atan2(other.get_imag(),other.get_real()
```